### DISCUSSION TITLE
Basic beginner set-up and configuration questions

### DISCUSSION THREAD

**Discussion by cswkim** [PaulRBerg/hardhat-template#98] [Q&A]
Thank you so much for this template, it's made getting started a great deal easier, especially for a complete novice to web3. I just had some really basic development workflow questions, my apologies if this is in a FAQ somewhere.

1) Is the general best practice to test your contract on the built-in `hardhat` network first right? Then a `testnet` like rinkby, ropsten, etc. And then finally when you are ready to `mainnet`?

2) You set the network in your `hardhat.config.ts`: `const config: HardhatUserConfig = { defaultNetwork: "hardhat",`...?

3) Is setting the network only relevant for deploying? Is anything else affected by changing the network?

4) If you follow the pattern in question 1), before deploying to each network do you need to manually change the network like in question 2) and then run `yarn deploy`? In general when you want to keep switching networks and run deployments, what is the best practice?

5) I set-up an Infura account and created a project and used it's `PROJECT ID` as my `.env` `INFURA_API_KEY`. Is Infura for deploying to testnets and mainnet?

6) You don't explicitly set the owner of the contract when deploying? I'm a little confused as to how the public contract methods with the `onlyOwner` modifier work? After I deployed to the default `hardhat` network and opened a console using `hardhat console --network hardhat` I could execute public `onlyOwner` functions. But what about contracts deployed to a testnet or mainnet? Generally if you want to run some public `onlyOwner` functions one-off, like let's say a `pause()` or `withdraw()` function, would you do this from your local machine using `hardhat console` even for mainnet?

I realize now most of the questions pertain to `hardhat` specifically, let me know and I'll just move them to a proper forum and close this discussion.

**Comment by PaulRBerg** [comment]
Hi @cswkim, glad you like the template. I will reply to your questions below but you might want to go through the [Hardhat tutorial](https://hardhat.org/tutorial/) and check out the [Hardhat documentation](https://hardhat.org/getting-started/). Also, you might want to join the [Hardhat Discord server](https://hardhat.org/discord) and ask the Hardhat maintainers questions from now on - I am not a maintainer myself.

1. Yes. As a rule of thumb when developing with Hardhat you always want to test your protocol against the [Hardhat Network](https://hardhat.org/hardhat-network/).
2. Yes.
3. It is relevant for every command you run with Hardhat.
4. No, you don't change the code, you just pass the name of the network via the `--network` CLI argument. Check out the Hardhat docs.
5. Infura is for deploying to whatever network they support.
6. This is outside the scope of this discussions channel, I recommend you ask your question in the [Ethereum StackExchange forum](https://ethereum.stackexchange.com/).
